wallet/src/key_chain/vrf_key_chain/mod.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key_chain::{KeyChainError, KeyChainResult};
+use common::address::Address;
+use common::chain::ChainConfig;
+use crypto::key::hdkd::child_number::ChildNumber;
+use crypto::key::hdkd::derivable::{Derivable, DerivationError};
+use crypto::key::hdkd::u31::U31;
+use crypto::vrf::{ExtendedVRFPublicKey, VRFPublicKey};
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use utils::const_value::ConstValue;
+use utils::ensure;
+use wallet_storage::{WalletStorageReadLocked, WalletStorageWriteLocked};
+use wallet_types::keys::KeychainUsageState;
+use wallet_types::AccountId;
+
+/// A child key hierarchy for an AccountKeyChain. This normally implements the receiving and change
+/// addresses key chains. It uses soft derivation to generate addresses (xpub).
+pub struct VrfKeySoftChain {
+    /// The specific chain this KeyChain is based on, this will affect the address format
+    chain_config: Arc<ChainConfig>,
+
+    /// The account id this leaf key chain belongs to
+    account_id: AccountId,
+
+    /// The parent key of this key chain
+    parent_pubkey: ConstValue<ExtendedVRFPublicKey>,
+
+    /// The parent key of this key chain
+    legacy_testnet_pubkey: ConstValue<ExtendedVRFPublicKey>,
+
+    /// The derived keys for the receiving funds or change. Those are derived as needed.
+    derived_public_keys: BTreeMap<ChildNumber, ExtendedVRFPublicKey>,
+
+    /// All the public key that this key chain holds
+    public_key_to_index: BTreeMap<VRFPublicKey, ChildNumber>,
+
+    /// The usage state of this key chain
+    usage_state: KeychainUsageState,
+}
+
+impl VrfKeySoftChain {
+    pub fn new_empty(
+        chain_config: Arc<ChainConfig>,
+        account_id: AccountId,
+        parent_pubkey: ExtendedVRFPublicKey,
+        legacy_testnet_pubkey: ExtendedVRFPublicKey,
+    ) -> Self {
+        Self {
+            chain_config,
+            account_id,
+            parent_pubkey: parent_pubkey.into(),
+            derived_public_keys: BTreeMap::new(),
+            public_key_to_index: BTreeMap::new(),
+            legacy_testnet_pubkey: legacy_testnet_pubkey.into(),
```

Comment:
It's not only for testnet. It can still (theoretically) be used for mainnet. The derivation mechanism is legacy though.